shared/src/main/scala/mlscript/ConstraintSolver.scala

```diff
@@ -332,6 +332,21 @@ class ConstraintSolver extends NormalForms { self: Typer =>
       trace(s"DNF DISCHARGE CONSTRAINTS") {
         lhsCons.foreach(c => rec(c._1, c._2, false))
       }()
+
+      // * finish when a type appears at both sides
+      val both = lhsCs.exists(s => s.lnf match {
+        case LhsTop => false
+        case LhsRefined(lbs, ltt, lr, ltr) =>
+          rhs.cs.exists(c => c.lnf match {
+            case LhsTop => false
+            case LhsRefined(rbs, rtt, rr, rtr) => 
+              !ltt.intersect(rtt).collect{ case sk: SkolemTag => sk }.isEmpty
```

Comment:
Why not like this? It's much more efficient.
```suggestion
              ltt.exists(rtt.contains)
```
Also I think this could be generalized to all abstract tags, not just skolem tags, no?

Reply:
maybe not? class tags have type members associated with and need to be compared?

Reply:
A class tag represents the type of a class instance whose type members are unknown. If you find the same one on both sides of a constraint, you can obviously finish the constraint right there.

Reply:
okay